### PR TITLE
Parse and trim InstallCode proposal payload

### PR DESCRIPTION
**Pull request by jasonz-dfinity**
# Motivation

The InstallCode proposal can be large, and we want to avoid letting the client download the entire WASM to calculate the hash.

# Changes

Returns wasm hash and upgrade args hex and hash for InstallCode proposals

# Tests

* Unit tests

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

File: frontend/src/lib/i18n/en.governance.json [hunk 1]
```diff
@@ -104,7 +107,10 @@
     "Motion": "A motion is a text that can be adopted or rejected. No code is executed when a motion is adopted. An adopted motion should guide the future strategy of the Internet Computer ecosystem.",
     "SetSnsTokenSwapOpenTimeWindow": "Start decentralization swap.",
     "OpenSnsTokenSwap": "Open decentralization swap.",
-    "CreateServiceNervousSystem": "Create a new Service Nervous System (SNS)."
+    "CreateServiceNervousSystem": "Create a new Service Nervous System (SNS).",
+    "InstallCode": "Install code on a canister.",
+    "StopOrStartCanister": "Stop or start a canister.",
+    "UpdateCanisterSettings": "Update settings of a canister."
```
Comment (mstrasinskis) [inline frontend/src/lib/i18n/en.governance.json:113]: To avoid repetitive text, maybe comments from the IC repo could be handy ([example](https://github.com/dfinity/ic/blob/3909a2cfe50518d54d217d4aa5918578591beb9c/rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs#L654))?

Comment (jasonz-dfinity) [reply frontend/src/lib/i18n/en.governance.json]: Done. I based on the comments but changed a little.

**Review by mstrasinskis — APPROVED** [review]
LGTM, thank you!

File: frontend/src/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.svelte [hunk 2]
```diff
@@ -30,7 +30,6 @@
   }
 </script>
 
-<!-- Only proposals with nnsFunctionKey and proposalId have payload -->
-{#if nnsFunctionKey !== undefined && proposalId !== undefined}
+{#if shouldDisplayPayload && proposalId !== undefined}
```
Comment (dskloetd) [inline frontend/src/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.svelte:33]: Does this actually prevent loading the payload or does it still load the payload but just not display it?

File: frontend/src/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.svelte
Comment (dskloetd) [inline frontend/src/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.svelte:1]: This file also needs unit tests.

File: rs/proposals/src/def.rs
Comment (dskloetd) [inline rs/proposals/src/def.rs:1]: Does the backend changes and frontend changes have to be in the same PR? Do they depend on each other?

File: rs/proposals/src/def.rs [hunk 3]
```diff
@@ -505,3 +511,22 @@ pub struct SubnetRentalRequest {
...
+
+impl From<&InstallCode> for InstallCodeTrimmed {
+    fn from(install_code: &InstallCode) -> Self {
+        InstallCodeTrimmed {
+            wasm_module_hash: install_code
+                .wasm_module
+                .as_ref()
+                .map(|wasm_module| calculate_hash_string(wasm_module))
+                .unwrap_or_default(),
+            arg_hex: install_code.arg.as_ref().map(hex::encode).unwrap_or_default(),
```
Comment (dskloetd) [inline rs/proposals/src/def.rs:529]: Currently we also display args as JSON. Have we made product decision not to do that anymore?

File: rs/proposals/src/lib.rs [hunk 4]
```diff
@@ -114,14 +114,27 @@ fn decode_arg(arg: &[u8], arg_types: &IDLTypes) -> String {
...
-        transform_payload_to_json(f.nns_function, &f.payload).unwrap_or_else(|e| {
-            let error_msg = "Unable to deserialize payload";
-            serde_json::to_string(&format!("{error_msg}: {e:.400}")).unwrap_or_else(|_| format!("\"{error_msg}\""))
-        })
-    } else {
-        serde_json::to_string("Proposal has no payload")
-            .unwrap_or_else(|err| unreachable!("Surely a fixed string can be serialized as JSON?  Err: {err:?}"))
+    let action = proposal_info.proposal.as_ref().and_then(|p| p.action.as_ref());
+    process_proposal_action(action)
+        .unwrap_or_else(|error_message| serde_json::to_string(&error_message).unwrap_or(error_message))
```
Comment (dskloetd) [inline rs/proposals/src/lib.rs:119]: I find the changes here a bit hard to follow. It's unclear to me what are functional changes vs. just syntax changes.
It looks like we now have one function just calling another function. Does the extra layer provide some value or is it necessary for something.
And then there is the change from `if` to `match`.
And the change of error message.
Can you try to keep the structure as much as same as possible so it's easier to review and explain the changes that are actually necessary?

File: rs/proposals/src/tests/mod.rs [hunk 5]
```diff
@@ -1,15 +1,42 @@
...
 }
+
+#[test]
+fn process_proposal_payload_install_code() {
+    let install_code_action = Action::InstallCode(InstallCode {
+        canister_id: Some(Principal::from_text("rrkah-fqaaa-aaaaa-aaaaq-cai").unwrap()),
+        wasm_module: Some(vec![1, 2, 3, 4].into()),
+        arg: Some(vec![5, 6, 7, 8].into()),
+        skip_stopping_before_installing: Some(false),
+        install_mode: Some(1),
```
Comment (dskloetd) [inline rs/proposals/src/tests/mod.rs:31]: Why are `skip_stopping_before_installing` and `install_mode` not rendered?

File: frontend/src/tests/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.spec.ts [hunk 6]
```diff
@@ -32,32 +29,108 @@ describe("NnsProposalProposerPayloadEntry", () => {
...
+      expect(nnsDappMock.getProposalPayload).toBeCalledTimes(0);
+      render(NnsProposalProposerPayloadEntry, {
+        props: {
+          proposal: proposalWithInstallCodeAction,
+          proposalId: mockProposalInfo.id,
+        },
+      });
+
+      await runResolvedPromises();
+      expect(nnsDappMock.getProposalPayload).toBeCalledTimes(1);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.spec.ts:86]: I know it wasn't there before but it would be good to also expect the params:
```
expect(nnsDappMock.getProposalPayload).toBeCalledWith({
  ...
});
```

File: rs/proposals/src/def.rs
Comment (dskloetd) [reply rs/proposals/src/def.rs:1]: This PR is on the big side according to our standards. I would prefer if the backend change was a separate PR that can be merged first. Thanks!

File: rs/proposals/src/tests/mod.rs (hunk 5 above)
Comment (dskloetd) [reply rs/proposals/src/tests/mod.rs:58]: Are they still rendered in the UI though? If not via `process_proposal_payload`, then how?
I feel like I'm missing something obvious.

File: rs/proposals/src/lib.rs (hunk 4 above)
Comment (dskloetd) [reply rs/proposals/src/lib.rs]: Thanks. If you think the tests are easier to review after the refactoring. Feel free to do the refactoring first and then rebase.

File: rs/proposals/src/tests/mod.rs (hunk 5 above)
Comment (jasonz-dfinity) [reply rs/proposals/src/tests/mod.rs:58]: The rest are rendered within the `InstallCode` section returned by clients calling NNS Governance directly (parsed by ic-js)

File: rs/proposals/src/def.rs
Comment (jasonz-dfinity) [reply rs/proposals/src/def.rs:1]: Done!

File: frontend/src/tests/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.spec.ts (hunk 6 above)
Comment (jasonz-dfinity) [reply frontend/src/tests/lib/components/proposal-detail/NnsProposalProposerPayloadEntry.spec.ts]: Fixing it in the next PR.

**Comment by dskloetd** [comment]
I realized I was confused about what does and doesn't get loaded by the frontend and what gets rendered by the backend.
So I did a bit of digging to clear up my confusion.

I found that while you can pass any proposal ID to `get_proposal_payload`, it really only works for `ExecuteNnsFunction` actions because only those proposals actually have a `payload` [field](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/canister/governance.did#L164C58-L164C65) and it's that field that is rendered by `get_proposal_payload`. But the InstallCode action doesn't have a `payload` field so calling `get_proposal_payload` for it doesn't really make sense. First I thought you were treating the `wasm_module` as the payload, but that's not quite true either because data derived from the `arg` field are also rendered in the response.

Then I was also confused about what gets loaded in the FE and I realized that because of the `omit_large_fields` flag, the `payload` of the `ExecuteNnsFunction` action as well as the `wasm_module` and `arg` of the `InstallCode` action [are omitted](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/src/governance.rs#L969-L981). (Although note that the [documentation](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs#L3042C49-L3043C79) of the field says "Currently only omits the `logo` and `token_logo` field of CreateServiceNervousSystem proposals.") This made me wonder why the `arg` is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?

Considering all this, it seems wrong to use the `get_proposal_payload` method in this way. I can think of a few (mutually exclusive) options:
1. Actually add a payload field to the InstallCode action, which holds the wasm_module and arg fields.
2. Add a separate method instead of reusing get_proposal_payload.
3. Rename get_proposal_payload to something more generic (`render_omitted_large_fields`?).
4. Accept that it's not great but go with the "hack" anyway.

Maybe you can think of something else entirely? What do you think?

**Comment by dskloetd** [comment]
Do I also see correctly that the payload is omitted, even if the `omit_large_fields` field is not set?
https://github.com/dfinity/ic/blob/9320a958fa05cd25f895cc034869b22d6b01aee9/rs/nns/governance/src/governance.rs#L3668-L3683

I also found that it's not omitted, even when the field is set, if the payload is < 1000 bytes.
All these special cases make it really difficult to understand what's going on.

**Comment by jasonz-dfinity** [comment]
> I realized I was confused about what does and doesn't get loaded by the frontend and what gets rendered by the backend. So I did a bit of digging to clear up my confusion.
> 
> I found that while you can pass any proposal ID to `get_proposal_payload`, it really only works for `ExecuteNnsFunction` actions because only those proposals actually have a `payload` [field](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/canister/governance.did#L164C58-L164C65) and it's that field that is rendered by `get_proposal_payload`. But the InstallCode action doesn't have a `payload` field so calling `get_proposal_payload` for it doesn't really make sense. First I thought you were treating the `wasm_module` as the payload, but that's not quite true either because data derived from the `arg` field are also rendered in the response.
> 
> Then I was also confused about what gets loaded in the FE and I realized that because of the `omit_large_fields` flag, the `payload` of the `ExecuteNnsFunction` action as well as the `wasm_module` and `arg` of the `InstallCode` action [are omitted](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/src/governance.rs#L969-L981). (Although note that the [documentation](https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/api/src/ic_nns_governance.pb.v1.rs#L3042C49-L3043C79) of the field says "Currently only omits the `logo` and `token_logo` field of CreateServiceNervousSystem proposals.") This made me wonder why the `arg` is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?
> 
> Considering all this, it seems wrong to use the `get_proposal_payload` method in this way. I can think of a few (mutually exclusive) options:
> 
> 1. Actually add a payload field to the InstallCode action, which holds the wasm_module and arg fields.
> 2. Add a separate method instead of reusing get_proposal_payload.
> 3. Rename get_proposal_payload to something more generic (`render_omitted_large_fields`?).
> 4. Accept that it's not great but go with the "hack" anyway.
> 
> Maybe you can think of something else entirely? What do you think?

I agree, but I think the complexity comes from the NNS Governance, and there doesn't seem to be a good way of solving it in the NNS Dapp. Here's a prototype of the ideal solution I have in mind: https://github.com/dfinity/ic/pull/822 which avoids returning large fields in the `InstallCode` entirely. However, it is a prototype as well as a breaking change, so I don't expect it to be merged very soon. I hope we can treat the current solution as a hack and accept it. WDYT?

**Comment by jasonz-dfinity** [comment]
> Do I also see correctly that the payload is omitted, even if the `omit_large_fields` field is not set? https://github.com/dfinity/ic/blob/9320a958fa05cd25f895cc034869b22d6b01aee9/rs/nns/governance/src/governance.rs#L3668-L3683
> 
> I also found that it's not omitted, even when the field is set, if the payload is < 1000 bytes. All these special cases make it really difficult to understand what's going on.

You are right, I also spent a lot of time trying to understand it...

File: rs/proposals/src/lib.rs [hunk 7]
```diff
@@ -114,14 +114,26 @@ fn decode_arg(arg: &[u8], arg_types: &IDLTypes) -> String {
...
+            })
+        }
+        Some(Action::InstallCode(install_code)) => {
+            let InstallCodeTrimmed {
+                wasm_module_hash,
+                arg_hex,
+                arg_hash,
+            } = InstallCodeTrimmed::from(install_code);
+            format!(
+                "{{\"wasm_module_hash\":\"{wasm_module_hash}\",\"arg_hex\":\"{arg_hex}\",\"arg_hash\":\"{arg_hash}\"}}"
```
Comment (dskloetd) [inline rs/proposals/src/lib.rs:132]: Can we use a standardized way to create the JSON from a struct instead of hand-crafting a string?

**Comment by dskloetd** [comment]
My main concern was that right now we can still change the `InstallCode` type because it's not used yet (right?) and doing it later will be more difficult.

Also, you didn't address this:

> This made me wonder why the arg is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?

**Review by dskloetd — APPROVED** [review]

**Comment by jasonz-dfinity** [comment]
> My main concern was that right now we can still change the `InstallCode` type because it's not used yet (right?) and doing it later will be more difficult.
> 
> Also, you didn't address this:
> 
> > This made me wonder why the arg is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?

Yes, it would be nice to make the improvements before it is enabled. However, it is also difficult to change the timeline for the new topics that are already communicated. I think a good strategy is to have the current implementation as the tentative plan, and try to push through the ideal solution before 

> My main concern was that right now we can still change the `InstallCode` type because it's not used yet (right?) and doing it later will be more difficult.
> 
> Also, you didn't address this:
> 
> > This made me wonder why the arg is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?

> My main concern was that right now we can still change the `InstallCode` type because it's not used yet (right?) and doing it later will be more difficult.

I agree, but it's not good to let the large fields improvement to block the topic split. I'll try to roll out the improvement before enabling the proposal type, but it will be best-effort.

> This made me wonder why the arg is removed as well. When is it expected to be large? How large? And should we really include the full args in the response when it is large?

I think they are not large, realistically. If they are indeed large, the current behavior with ExecuteNnsFunction is that, with list_proposals they are blank, and when displaying a single proposal, they are still rendered through the `get_proposal_payload` path no matter how large they are. I think the current implementation preserves that behavior and add the hash on top of it. By having both of the hash and the hex, it will be possible for the proposers to change their verification instruction from verifying hex to hash, and if everyone changes to verifying hash, we can then remove the hex display.

File: rs/proposals/src/lib.rs (hunk 7 above)
Comment (jasonz-dfinity) [reply rs/proposals/src/lib.rs]: Done. Using `serde_json::json!` macro.

File: rs/proposals/src/lib.rs [hunk 8]
```diff
@@ -128,9 +129,12 @@ pub fn process_proposal_payload(proposal_info: &ProposalInfo) -> Json {
...
             } = InstallCodeTrimmed::from(install_code);
-            format!(
-                "{{\"wasm_module_hash\":\"{wasm_module_hash}\",\"arg_hex\":\"{arg_hex}\",\"arg_hash\":\"{arg_hash}\"}}"
-            )
+            json!({
+                "wasm_module_hash": wasm_module_hash,
+                "arg_hex": arg_hex,
+                "arg_hash": arg_hash,
+            })
+            .to_string()
```
Comment (dskloetd) [inline rs/proposals/src/lib.rs:137]: Can't you just do
```
let trimmed = InstallCodeTrimmed::from(install_code);
json!(trimmed).to_string()
```
?

**Review by dskloetd — APPROVED** [review]

**Comment by dskloetd** [comment]
>> why the arg is removed as well

> I think they are not large

If the `arg` is not large, then why is it removed in `omit_large_fields`?
https://github.com/dfinity/ic/blob/c0373c673f8e8cc95ea6e9cdcf2434b789243e19/rs/nns/governance/src/governance.rs#L979

File: rs/proposals/src/lib.rs (hunk 8 above)
Comment (jasonz-dfinity) [reply rs/proposals/src/lib.rs]: Nice... I didn't know that. Done.
